### ISSUE TITLE
Catalogue Clean - Improved playlist behaviour

### ISSUE THREAD

**Issue by RavenRabbit** [ampache/ampache#1745]
# Environment

 * Ampache version: 3.8.8-release
 * Web server + version:  Apache/2.4.25
 * Server operating system: Debian Stretch

 * Client type:
   - Web
   - Ampache

# Description

In its current implementation songs get removed from all playlists if the catalogue is cleaned and the exact file is not present on the filesystem anymore (e.g., it has been renamed or actually deleted). However, for users who have time-consumingly created a playlist, this might result in empty or incomplete playlists even if the same song is available from a different album or has only been renamed when correcting a typo. The issue here is that the songs are just gone and a user can't even check which files are gone (except from manually comparing to an exported playlist).
 
Therefore I'd like to suggest a few improvements:
 - When 'cleaning' the catalogue, mark the entries as `deleted` for the playlists.
 	- Retain only basic information about the song (like title, artist, album, length) and only for the playlist view (maybe create this in a different table?)
 	- Remove these entries once there aren't any references to it in any playlist anymore.
 - In the playlist view mark songs as deleted using a red background or a strikethrough text-decoration and a mouse over text stating that the song is unavailable and has been moved/renamed on the server
 - During playback just skip the files? Maybe apply the same formatting as in the playlist view.
 - In the playlist view at the top offer a `Remove unavailable files`. As admins can see all playlists in the playlist view, this should be available for them for all playlists.
 - Introduce a setting to enable/disable this on a server/user/group basis
 - Optionally: add button for unavailable files to directly search for the same title/artist combination in the catalogue to speed-up replacing it with a different version.

This would allow users to have better control over removed files on the server. User and admins that don't care can configure this in the settings to keep the current behaviour of removing the files without any further notice.
Users who only want to be informed about removed files, but don't bother replacing them can just clean their playlist(s) using the button on the top.
Users who want to replace the files can do so as well.

I think this would certainly improve the bahaviour of playlists for catalogues that replace/rename files from time to time and results in a better user experience than just magically disappearing files from playlists.

**Comment by lachlan-00** [comment]
The way to do this would be to just update tags on the file and then rename using ampache. By doing it outside of ampache it's not easy to follow changes.

If you rename in ampache it will make all the changes without any issue.

I'm going to work on extending the renaming to better match average naming. (Ampache doesn't support my library naming.) but this isn't really something i think would work.

**Comment by RavenRabbit** [comment]
Cleaning the catalogue discovers these kind of changes as it's removing tracks from the database and all playlists if it's not available on the filesystem anymore. The idea was to mark them as removed and displaying them as deleted in the playlist view instead of making them disappear without a trace.
So it's less about following filename changes on disk, which is indeed a hard task, but giving to keeping the entries in a playlist sound and allowing the user to repair it.

In my case the music is mounted read-only so renaming from Ampache is not an option and not necessary for this feature.

**Comment by lachlan-00** [comment]
If you don't rename the file the tag changes are followed.

What I'm saying is that of you delete a file ampache doesn't care why. It's deleted.

**Comment by lachlan-00** [comment]
Sorry swiping on mobile.
What I'm saying is that the files on Ampache should be static and the tags are the only thing Ampache actually worries about.

**Comment by RavenRabbit** [comment]
Yes, files _should_ be static. However, the files are located on a disk, often in some sort of artist/album/ folder structure. Renaming the file and or path either to fix spelling errors (they also happen when tagging from sources like Musicbrainz, as tags are changing on there as well), or to differentiate albums if there's some sort of collision (happens from time to time). So there's no need to view filenames as they are changing at constant rate, but they can and will change.
And I think it doesn't make a lot of sense to fix a spelling error in the tag and not update the filename. I see that this could then maybe be done through Ampache, but I don't think Ampache can be seen as a primary music tagging software when there are tools like beets, picard, mp3tagger,... out there.

> What I'm saying is that of you delete a file ampache doesn't care why. It's deleted.
My concern is only about _playlists_, as on instances with more than one user the renamed files would simply disappear without a trace. Playlists are generated by the users, and removing entries from there is not a nice thing to do, as it makes for a bad experience ("I definitely put that song with the name I can't remember anymore on there, where is it?").
Looking at applications like iTunes: When a file disappears from the disk (or a filename changes) the file is marked with a white-on-grey exclamation mark to indicate that the file is missing. The software then allows you to provide the relocated path or delete it. This is in my opinion a desirable behaviour for the playlists (not the library as a whole, just the playlists).

**Comment by kuzi-moto** [comment]
@lachlan-00 Do playlists use mbids at all? Seems like if all you do is change a filename, that Ampache should recognize that it's the same song and continue working in the playlist. If there is no mbid in the song, then it would be impossible for Ampache to know that, as there could be several songs with the exact same file name.

However, for playlists I'd say they are a special case. I think that some people spend a lot of time curating playlists, so knowing what was in the playlist can be more important than the music data itself. I know people who use Spotify get pretty disappointed when music from their playlist just disappears with no notice.

Maybe if it's too much trouble to change how playlists work (as in marking a song as deleted vs. just removing it), then we can send any user with that song in a playlist that it was deleted? I'm not sure if Ampache has a method of sending 'system' messages to users or not though.

**Comment by lachlan-00** [comment]
The easy way to bypass all this is to use the bin/sort_files.inc script.
It needs updating to extend the values it matches but it will manage all the changes and keep the original song id's without deleting them.

If ampache isn't your point of truth the issue is that ampache catalog updates are done one song at a time. To alleviate this it would be better to disable songs in the song table and then update them if the new file matches.

I'm not sure if disabling the song will remove from a playlist as well but it would be a better option for less work than changing the playlist table.
